### PR TITLE
reapply custom miraheze changes

### PR DESCRIPTION
**Pull request by RhinosF1**
Per https://github.com/wikimedia/mediawiki/compare/REL1_35...miraheze:REL1_35

The only part I can't just apply is https://github.com/miraheze/mediawiki/blob/8a58e08361b6ea9cd22b9bd56b7a5de7d08343d1/includes/clientpool/SquidPurgeClient.php#L196-L235 as that file no longer exists

### CODE REVIEW

File: includes/deferred/CdnCacheUpdate.php [hunk 1]
```diff
@@ -295,24 +295,29 @@ private static function naivePurge( array $urls ) {
...
-			foreach ( $wgCdnServers as $server ) {
-				$reqs[] = ( $baseReq + [ 'proxy' => $server ] );
+			// Southparkfan hack start (adds x-device)
+			foreach ( [ 'desktop', 'phone-tablet' ] as $deviceHeader ) {
+                $url = self::expand( $url );
+                $urlInfo = wfParseUrl( $url );
+				$urlHost = strlen( $urlInfo['port'] ?? null )
+					? IP::combineHostAndPort( $urlInfo['host'], $urlInfo['port'] )
+					: $urlInfo['host'];
+				$baseReq = [
```
Comment (Southparkfan) [inline includes/deferred/CdnCacheUpdate.php:305]: indent

File: includes/deferred/CdnCacheUpdate.php [hunk 2]
```diff
@@ -295,24 +295,29 @@ private static function naivePurge( array $urls ) {
...
-				'headers' => [
-					'Host' => $urlHost,
-					'Connection' => 'Keep-Alive',
-					'Proxy-Connection' => 'Keep-Alive',
-					'User-Agent' => 'MediaWiki/' . MW_VERSION . ' ' . __CLASS__
-				]
-			];
-			foreach ( $wgCdnServers as $server ) {
-				$reqs[] = ( $baseReq + [ 'proxy' => $server ] );
+			// Southparkfan hack start (adds x-device)
```
Comment (Southparkfan) [inline includes/deferred/CdnCacheUpdate.php:298]: Can we include documentation: **why** do we include an X-Device header on each purge?

**Review by Southparkfan — CHANGES_REQUESTED** [review]

File: includes/deferred/CdnCacheUpdate.php (hunk 2 above)
Comment (RhinosF1) [reply includes/deferred/CdnCacheUpdate.php:298]: We can but they are your hacks. Can you recall why?

Comment (Southparkfan) [reply includes/deferred/CdnCacheUpdate.php:298]: > In Varnish, we store two variants of an article: the mobile layout (via MobileFrontend) and the desktop version. This is done by running vcl_hash over the X-Device header. The X-Device header is set at Varnish level and can be one of ['mobile-tablet', 'desktop']. However, when purging a resource using merely the URL path, it's not possible to purge both the mobile and desktop variants. In order to purge properly, MediaWiki must set each possible X-Device value in the X-Device header, one per PURGE request.

**Comment by Reception123** [comment]
@RhinosF1 Since it's your task, do you want to add the comment left by SPF to the file?

**Comment by RhinosF1** [comment]
Did it mobile @Reception123 but there you go! Feel free to tidy up.

**Review by Southparkfan — APPROVED** [review]

**Pull request merged**